gemini/use-cases/retrieval-augmented-generation/developer_code_chat/utils/intent_routing.py

```diff
@@ -0,0 +1,401 @@
+        self, model, text, enabled_programming_language
+    ):
+        """Extract Programming lanuages of incoming message"""
+
+        response = model.generate_content(
+            f"""
+            You are strict programming languages extractor.
+
+            Instructions:
+            1. Extract only programming languages from message.
+            2. Don't return any other languages other than programming.
+            3. return [] if no programming lanuage in mentioned in message.
+            4. {enabled_programming_language} these are the programming languages.
+
+            Examples:
+            write a code for fibonacci series in C++? : ["C++"]
+            write a code using C# to generate palindrome series : ["C#"]
+            using python, write a sample application code to create endpoint? : ["Python"]
+            what are classes in Java? : ["Java"]
+            write a code for reverse a string : []
+            what are data types? : []
+
+            What are the programming languages mentioned in below message?
+            MESSAGE:{text}
+            programming languages:""",  # pylint: disable=C0301:line-too-long
+            generation_config=self.genai_qna_parameters,
+        )
+
+        programming_lang = response.text
+        program_lang_in_query = []
+        if programming_lang:
+            try:
+                programming_lang = programming_lang.replace("'", '"')
+                # print(programming_lang)
+                programming_lang = json.loads(programming_lang)
+                program_lang_in_query = [
+                    x.lower().replace(" ", "").strip() for x in programming_lang
+                ]
+            except Exception:  # pylint: disable=W0718,W0703
+                self.logger.info("Error while extracting programming language.")
+        return program_lang_in_query
+
+    def check_programming_language_in_query(self, model, text, intent):
+        """Check if which programming language is mentioned in the use query"""
+
+        if intent == "WRITE_CODE":
+            enabled_programming_language = self.config["default"][
+                "enabled_programming_language"
+            ]
+        else:
+            enabled_programming_language = self.config["default"][
+                "enabled_qna_programming_language"
+            ]
+        enabled_programming_language = enabled_programming_language.split(",")
+        enabled_programming_language_list = [
+            x.lower().replace(" ", "").strip() for x in enabled_programming_language
+        ]
+        program_lang_in_query = self.get_programming_lanuage_from_query(
+            model, text, enabled_programming_language
+        )
+        allowed_language_in_query = set(enabled_programming_language_list).intersection(
+            set(program_lang_in_query)
+        )
+
+        return program_lang_in_query, allowed_language_in_query
+
+    def classify_intent(self, text, session_state, model, chat_model, genai_qna):
```

Comment:
This function is very heavily nested and difficult to follow, can it be simplified?